lib/Transforms/TensorToScalars/TensorToScalars.cpp

```diff
@@ -0,0 +1,144 @@
+                                                ValueRange inputs,
+                                                Location loc) {
+  return builder.create<tensor::FromElementsOp>(loc, resultType, inputs);
+}
+
+static std::optional<SmallVector<Value>> buildExtractOps(OpBuilder &builder,
+                                                         TypeRange resultTypes,
+                                                         Value input,
+                                                         Location loc) {
+  // This conversion only operates on tensors of static shape
+  RankedTensorType inputType = dyn_cast<RankedTensorType>(input.getType());
+  if (!inputType || !inputType.hasStaticShape()) return {};
+
+  // Create extract ops in "natural" order (dimension-by-dimension)
+  SmallVector<Value> values;
+  for (auto dim : inputType.getShape()) {
+    for (int i = 0; i < dim; ++i) {
+      Value index = builder.create<arith::ConstantIndexOp>(loc, i);
+      Value element = builder.create<tensor::ExtractOp>(loc, input, index);
+      values.push_back(element);
+    }
+  }
+  return values;
+}
+
+class ConvertFromElementsOp
+    : public OneToNOpConversionPattern<tensor::FromElementsOp> {
+ public:
+  using OneToNOpConversionPattern<
+      tensor::FromElementsOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::FromElementsOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    // This conversion only operates on tensors of static shape,
+    // but no check is necessary here as from_elements' shape is always static
+
+    // Replace the current op with the flattened operands.
+    // This should already match the "natural" order expected by this pass.
+    rewriter.replaceOp(op, adaptor.getFlatOperands(),
+                       adaptor.getResultMapping());
+    return success();
+  }
+};
+
+class ConvertInsertOp : public OneToNOpConversionPattern<tensor::InsertOp> {
+ public:
+  using OneToNOpConversionPattern<tensor::InsertOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::InsertOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    // This conversion only operates on tensors of static shape
+    if (!op.getResult().getType().hasStaticShape()) return failure();
+
+    // We can only support statically known indices
+    // that have been constant-folded to a single arith.constant op
+    for (auto idx : op.getIndices()) {
+      if (!llvm::isa<arith::ConstantIndexOp>(idx.getDefiningOp()))
+        return failure();
+    }
+
+    // Compute the insertion offset (in dimension-by-dimension order):
+    int64_t multiplier = 1;
+    int64_t offset = 0;
+    for (auto [dim, idx] :
+         llvm::zip(op.getResult().getType().getShape(), op.getIndices())) {
+      offset +=
+          idx.getDefiningOp<arith::ConstantIndexOp>().value() * multiplier;
```

Comment:
instead of looping twice you may as well check for the failure condition here. IIUC, `dyn_cast` is relatively expensive in MLIR.